File: account/src/main/java/dev/exit/asset/account/AccountController.java
package dev.exit.asset.account;

import java.util.List;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import dev.exit.asset.common.CurrentUserProvider;
import dev.exit.asset.account.AccountKind;
import dev.exit.asset.account.AssetAccount;
import dev.exit.asset.account.AssetAccountRepository;

@RestController
@RequestMapping("/api/v1/accounts")
public class AccountController {

	private final CurrentUserProvider currentUser;

	private final AssetAccountRepository accounts;

	public AccountController(CurrentUserProvider currentUser, AssetAccountRepository accounts) {
		this.currentUser = currentUser;
		this.accounts = accounts;
	}

	@GetMapping
	public List<AssetAccount> list() {
		return accounts.findByOwnerIdOrderByCreatedAtDesc(currentUser.ownerId());
	}

	@PostMapping
	public AssetAccount create(@Valid @RequestBody CreateAccountRequest request) {
		return accounts.save(new AssetAccount(currentUser.ownerId(), request.name(), request.kind(), request.institution()));
	}

	public record CreateAccountRequest(@NotBlank String name, @NotNull AccountKind kind, @NotBlank String institution) {
	}
}

File: account/src/main/kotlin/dev/exit/asset/account/AccountController.kt
package dev.exit.asset.account

import dev.exit.asset.common.CurrentUserProvider
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/accounts")
class AccountController(
    private val currentUser: CurrentUserProvider,
    private val accounts: AssetAccountRepository,
) {

    @GetMapping
    fun list(): List<AssetAccount> = accounts.findByOwnerIdOrderByCreatedAtDesc(currentUser.ownerId())

    @PostMapping
    fun create(@Valid @RequestBody request: CreateAccountRequest): AssetAccount =
        accounts.save(AssetAccount(currentUser.ownerId(), request.name, request.kind, request.institution))

    data class CreateAccountRequest(
        @field:NotBlank val name: String,
        @field:NotNull val kind: AccountKind,
        @field:NotBlank val institution: String,
    )
}
